runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -132,16 +208,56 @@ public void onTimer(
                 + "since window is too far behind inputWatermark:{}",
             timestamp,
             window,
-            cleanupTimer.currentInputWatermarkTime());
+            stepContext.timerInternals().currentInputWatermarkTime());
       } else {
         doFnRunner.onTimer(timerId, window, timestamp, timeDomain);
       }
     }
   }
 
-  @Override
-  public void finishBundle() {
-    doFnRunner.finishBundle();
+  // this needs to be optimized (Sorted Map State)
```

Comment:
Is it a performace implication worth to mention to annotation users? 

Can it be link it to a JIRA?

Reply:
Actually, it turns out that when the buffer is flushed not per single element, but for each input watermark move, this optimization might be negligible. I'm seeing good performance even without it. JIRA would be good, as when there is Sorted Map State, then this can be leveraged.

Reply:
I will create the JIRA after this is merged.

Reply:
Created https://issues.apache.org/jira/browse/BEAM-9256